primefield/src/monty.rs

```diff
@@ -0,0 +1,444 @@
+
+        Self::from_uint_unchecked(&Uint::from_u64(w))
+    }
+
+    /// Returns the bytestring encoding of this field element.
+    pub fn to_bytes(self) -> Array<u8, MOD::ByteSize>
+    where
+        Uint<LIMBS>: ArrayEncoding,
+    {
+        let mut repr = Array::<u8, MOD::ByteSize>::default();
+        debug_assert!(repr.len() <= MOD::ByteSize::USIZE);
+
+        let offset = MOD::ByteSize::USIZE.saturating_sub(repr.len());
+
+        match MOD::BYTE_ORDER {
+            ByteOrder::BigEndian => {
+                let padded = self.0.retrieve().to_be_byte_array();
+                repr.copy_from_slice(&padded[offset..]);
+            }
+            ByteOrder::LittleEndian => {
+                let padded = self.0.retrieve().to_le_byte_array();
+                repr.copy_from_slice(&padded[..offset]);
+            }
+        }
+
+        repr
+    }
+
+    /// Determine if this field element is odd: `self mod 2 == 1`.
+    ///
+    /// # Returns
+    ///
+    /// If odd, return `Choice(1)`.  Otherwise, return `Choice(0)`.
+    #[inline]
+    pub fn is_odd(&self) -> Choice {
+        self.0.retrieve().is_odd()
+    }
+
+    /// Determine if this field element is even: `self mod 2 == 0`.
+    ///
+    /// # Returns
+    ///
+    /// If even, return `Choice(1)`.  Otherwise, return `Choice(0)`.
+    #[inline]
+    pub fn is_even(&self) -> Choice {
+        !self.is_odd()
+    }
+
+    /// Determine if this field element is zero.
+    ///
+    /// # Returns
+    ///
+    /// If zero, return `Choice(1)`.  Otherwise, return `Choice(0)`.
+    #[inline]
+    pub fn is_zero(&self) -> Choice {
+        self.ct_eq(&Self::ZERO)
+    }
+
+    /// Translate field element out of the Montgomery domain, returning a [`Uint`] in canonical form.
+    #[inline]
+    pub const fn to_canonical(self) -> Uint<LIMBS> {
+        self.0.retrieve()
+    }
+
+    /// Add elements.
+    #[inline]
+    pub const fn add(&self, rhs: &Self) -> Self {
+        Self(MontyForm::add(&self.0, &rhs.0))
+    }
+
+    /// Double element (add it to itself).
+    #[inline]
+    #[must_use]
+    pub const fn double(&self) -> Self {
+        self.add(self)
+    }
+
+    /// Subtract elements.
+    #[inline]
+    pub const fn sub(&self, rhs: &Self) -> Self {
+        Self(MontyForm::sub(&self.0, &rhs.0))
+    }
+
+    /// Multiply elements.
+    #[inline]
+    pub const fn multiply(&self, rhs: &Self) -> Self {
+        Self(MontyForm::mul(&self.0, &rhs.0))
+    }
+
+    /// Negate element.
+    #[inline]
+    pub const fn neg(&self) -> Self {
+        Self(MontyForm::neg(&self.0))
+    }
+
+    /// Compute modular square.
+    #[inline]
+    #[must_use]
+    pub const fn square(&self) -> Self {
+        self.multiply(self)
+    }
+
+    /// Compute field inversion: `1 / self`.
+    #[inline]
+    pub fn invert<const UNSAT_LIMBS: usize>(&self) -> CtOption<Self>
+    where
+        Odd<Uint<LIMBS>>: PrecomputeInverter<Inverter = SafeGcdInverter<LIMBS, UNSAT_LIMBS>, Output = Uint<LIMBS>>,
```

Comment:
This bound is atrocious but can hopefully go away with binary XGCD (RustCrypto/crypto-bigint#875)

Reply:
Simplified it to:

```rust
MontyForm<MOD, LIMBS>: Invert<Output = CtOption<MontyForm<MOD, LIMBS>>>
```